### ISSUE TITLE
Reproducing HuBERT-ECG Inference Results on PTB-XL

### ISSUE THREAD

**Issue by elston** [Edoar-do/HuBERT-ECG#13]
Dear Edoardo,

My name is Denis, and I am a bioinformatics researcher currently exploring the application of transformer models in ECG pathology detection. I attempted to reproduce the inference results on the PTB-XL dataset using your repository: https://github.com/Edoar-do/HuBERT-ECG.

I followed the instructions in the reproducibility section and utilized the pre-trained large model from this repository: https://huggingface.co/Edoardo-BS/HuBERT-ECG-SSL-Pretrained. However, the results I obtained were significantly lower than those reported in your work. My notebook with the code is available here: https://gist.github.com/elston/b54899209f4a74fedfb550ebdfb39205. I also ensured that all preprocessing steps were performed as specified.

I suspect that I might have missed critical steps, such as fine-tuning the model on PTB-XL. I would greatly appreciate it if you could provide guidance on potential reasons for the low performance and advise on the necessary steps to achieve results comparable to yours.

Thank you in advance for your assistance and for your work on HuBERT-ECG.

Best regards,
Denis

![Image](https://github.com/user-attachments/assets/aa103527-0bfa-48c5-a376-c46c3612b9b7)

**Comment by Edoar-do** [comment]
Hi @elston , it seems you're not reproducing the results in our paper.
It seems you're using the HuBERT-ECG fine-tuned on CardioLearning to make inference on PTB-XL. That is a scenario we never tested in our paper, so no comparison can be made.
Instead, in order to reproduce the results in our paper, you have to fine-tune the SSL pre-trained HuBERT-ECG on PTB-XL (training partition) and then make inference on the corresponding test set.

**Comment by Edoar-do** [comment]
Hi @elston , sorry for the previous reply (now deleted). I've had the occasion of taking a closer look at your code now.
It seems you attached a linear classifier atop the pre-trained model but didn't perform any finetuning, therefore the model head still has randomly initiliased weigths and, additionally, the backbone has pre-trained task-agnostic weigths instead of PTB-XL's specific weigths.

In order to reproduce our results, you need to load the the pre-trained backbone, create an instance of hubert-ecg for classification (which attaches a linear layer classifier on top of it), fine-tune the model on ptb-xl and, eventually, run inference.

Bash scripts should help you do that

Let me know if I got your point and if you have any more question

**Comment by elston** [comment]
Thank you very much for quick response and clarification! 
So, you mean that fine-tuning is required before inference and  repo does not contain the ready-to-go ECG classifier and final classifier layer will have randomly initialized weights, so I need to freeze all layers except the last one and fine-tune that on any ECG data related to my own task? However, the rest layers of your ECG-classifier is ready to go and no need to move weights there?

**Comment by Edoar-do** [comment]
1) Yes, the repos on hugging face do not contain fine-tuned models ready for inference. Since most datasets are publicly available, one can easily fine-tune the pre-trained models (those on Hugging Face) using the code in this github repo 
2) No, if you freeze all model layers except the newly attached classification head you're going to perform a linear evaluation, not a fine-tuning. To perform a fine-tuning you don't have to freeze any layer, but you can follow my `finetune.py` implementation for more details (just to mention, as you might have read in the paper, the only component that is frozen during fine-tuning is the convolutional embedder). 

To sum it up:
1. load the pre-trained model
2. create an instance of HubertForECGClassification (this magically attaches a randomly initialised linear classifier on top of the backbone)
3. freeze the conv embedder and fine-tune everything else
4. use the fine-tuned model for inference on the test set

**Comment by Edoar-do** [comment]
No more activity
